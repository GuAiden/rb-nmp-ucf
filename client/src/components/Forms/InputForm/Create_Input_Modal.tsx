import React, { useState } from 'react';
import {
  Button,
  Form,
  FormControl,
  InputGroup,
  Modal,
  Popover,
  OverlayTrigger,
} from 'react-bootstrap';
import { Input } from '../Input_Types';
import './index.css';

type CreateInputModalProps = {
  onAddInput: (userInputs: Input) => void;
  inputList: Input[];
};

const initialState = {
  output: false,
  conversion: false,
} as Input;

/**
 * @returns A modal component that opens to handle form input for the 'inputs' page
 */
const CreateInputModal: React.FunctionComponent<CreateInputModalProps> = ({
  onAddInput,
  inputList,
}: CreateInputModalProps) => {
  /**
   * showState -> responsible for opening and closing modal
   * inputState -> responsible for storing form inputs
   * isConversionState -> responsible for handling conversion checkbox state
   * isOutputState -> responsible for handling output checkbox state
   */
  const [show, setShow] = useState(false);
  const [input, setInput] = useState<Input>(initialState);
  const [isConversion, setConversion] = useState<boolean>(false);
  const [isOutput, setOutput] = useState<boolean>(false);

  // Modal open and close handlers
  const handleShow = (): void => {
    setShow(true);
  };

  const handleClose = (): void => {
    setShow(false);
    setConversion(false);
    setOutput(false);
    setInput(initialState);
  };

  const isValidInput = (): boolean => {
    // Validate channelNumber uniqueness
    if (
      inputList.some(
        (savedInput) => savedInput.channelNumber === input.channelNumber,
      )
    ) {
      return false;
    }

    // Validate existence of channel number, name and output
    // Might want to validate data integrity as well later
    if (
      typeof input.channelName === 'undefined' ||
      typeof input.channelNumber === 'undefined' ||
      typeof input.units === 'undefined'
    ) {
      return false;
    }
    return true;
  };

  // CreateInputModal add button handler
  const handleAddInput = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (!isValidInput()) {
      e.preventDefault();
      return;
    }

    onAddInput(input);
    setShow(false);
    setOutput(false);
    setConversion(false);
    setInput(initialState);
  };

  // onChange handlers for form inputs
  const onChannelNameChange = (
    e: Parameters<
      NonNullable<React.ComponentProps<typeof FormControl>['onChange']>
    >[0],
  ): void => {
    setInput({ ...input, channelName: e.currentTarget.value as string });
  };

  const onChannelNumberChange = (
    e: Parameters<
      NonNullable<React.ComponentProps<typeof FormControl>['onChange']>
    >[0],
  ): void => {
    setInput({ ...input, channelNumber: parseInt(e.currentTarget.value, 10) });
  };

  const onUnitsChange = (
    e: Parameters<
      NonNullable<React.ComponentProps<typeof FormControl>['onChange']>
    >[0],
  ): void => {
    setInput({ ...input, units: e.currentTarget.value as string });
  };

  const onXChange = (
    e: Parameters<
      NonNullable<React.ComponentProps<typeof FormControl>['onChange']>
    >[0],
  ): void => {
    setInput({ ...input, x: parseInt(e.currentTarget.value, 10) });
  };

  const onYChange = (
    e: Parameters<
      NonNullable<React.ComponentProps<typeof FormControl>['onChange']>
    >[0],
  ): void => {
    setInput({ ...input, y: parseInt(e.currentTarget.value, 10) });
  };

  const onOutputChange = (): void => {
    setOutput(!isOutput);
    setInput({ ...input, output: !isOutput });
  };

  const onConversionChange = (): void => {
    setConversion(!isConversion);
    setInput({ ...input, conversion: !isConversion });
  };

  // Popover to let user know about invalid input
  const invalidInputPopover = (
    <Popover id="popover-basic">
      <Popover.Body className="my-auto">
        Please enter a valid:
        <ul className="list-p-0">
          <li>
            <b>Name</b>
          </li>
          <li>
            <b>Number</b>
          </li>
          <li>
            <b>Ouput</b>
          </li>
        </ul>
      </Popover.Body>
    </Popover>
  );

  // Button for when user input is still invalid and untrustworthy
  const DangerAddButton: React.FunctionComponent = () => (
    <OverlayTrigger
      trigger="click"
      placement="left"
      overlay={invalidInputPopover}
    >
      <Button
        variant="danger"
        size="lg"
        onClick={(e): void => handleAddInput(e)}
        className="border rounded-0 modal-add-wrapper px-5"
        id="danger-add-button"
      >
        Add
      </Button>
    </OverlayTrigger>
  );

  // Button for when a successful input can be added
  const AddButton: React.FunctionComponent = () => (
    <Button
      variant="secondary"
      size="lg"
      onClick={(e): void => handleAddInput(e)}
      className="border rounded-0 modal-add-wrapper px-5"
      id="add-button"
    >
      Add
    </Button>
  );

  return (
    <React.Fragment>
      <div className="float-left">
        <Button
          className="rounded-0 px-4 py-2 button-text margin-left"
          variant="primary"
          onClick={handleShow}
        >
          Create new
        </Button>
      </div>

      <Modal show={show} onHide={handleClose} centered size="lg">
        <Modal.Body className="input-modal">
          <div className="container p-5">
            <form>
              {/* CHANNEL NAME AND NUMBER INPUTS */}
              <div className="row">
                <header className="modal-header-wrapper">
                  CREATE NEW INPUT
                </header>
              </div>
              <div className="row">
                <Form.Group
                  className="mt-3 text-light"
                  controlId="channelName.ControlInput"
                >
                  <Form.Label>Channel Name</Form.Label>
                  <Form.Control
                    type="text"
                    className="text-input-wrapper"
                    onChange={(e): void => onChannelNameChange(e)}
                  />
                </Form.Group>
                <Form.Group
                  className="mt-3 text-light"
                  controlId="channelNumber.ControlInput"
                >
                  <Form.Label>Channel Number</Form.Label>
                  <Form.Control
                    type="text"
                    className="text-input-wrapper"
                    onChange={(e): void => onChannelNumberChange(e)}
                  />
                </Form.Group>
              </div>
              {/* UNITS INPUTS WITH OUTPUT/CONVERSION CHECKBOXES */}
              <div className="row mt-4 justify-content-around">
                <div className="col-6 my-auto">
                  <InputGroup className="input-group-sm mb-3">
                    <InputGroup.Text className="input-group-text-wrapper">
                      Units:
                    </InputGroup.Text>
                    <FormControl
                      aria-label="Units input"
                      aria-describedby="inputGroup-sizing-sm"
                      className="text-input-wrapper"
                      onChange={(e): void => onUnitsChange(e)}
                    ></FormControl>
                  </InputGroup>
                </div>
                <div className="col-lg-5 my-auto">
                  <>
                    {['checkbox' as const].map((type) => (
                      <div key={`inline-${type}`} className="mb-3">
                        <Form.Check
                          inline
                          label="Output?"
                          name="group1"
                          type={type}
                          checked={isOutput}
                          onChange={(): void => onOutputChange()}
                          id={`inline-${type}-1`}
                        />
                        <Form.Check
                          inline
                          label="Conversion?"
                          name="group1"
                          type={type}
                          checked={isConversion}
                          onChange={(): void => onConversionChange()}
                          id={`inline-${type}-2`}
                        />
                      </div>
                    ))}
                  </>
                </div>
              </div>
              {/* X AND Y INPUTS */}
              {isOutput && isConversion && (
                <div className="row mt-4 justify-content-around mx-auto">
                  <div className="col-lg-5">
                    <InputGroup className="input-group-sm mb-3">
                      <InputGroup.Text className="input-group-text-wrapper">
                        X:
                      </InputGroup.Text>
                      <FormControl
                        aria-label="X input value"
                        aria-describedby="inputGroup-sizing-sm"
                        className="text-input-wrapper"
                        onChange={(e): void => onXChange(e)}
                      ></FormControl>
                    </InputGroup>
                  </div>
                  <div className="col-lg-5">
                    <InputGroup className="input-group-sm mb-3">
                      <InputGroup.Text className="input-group-text-wrapper">
                        Y:
                      </InputGroup.Text>
                      <FormControl
                        aria-label="Y input value"
                        aria-describedby="inputGroup-sizing-sm"
                        className="text-input-wrapper"
                        onChange={(e): void => onYChange(e)}
                      ></FormControl>
                    </InputGroup>
                  </div>
                </div>
              )}
              {/* ADD OR SAVE MODAL BUTTONS */}
              <div className="row justify-content-end mt-4">
                <div className="col-3">
                  {isValidInput() ? <AddButton /> : <DangerAddButton />}
                </div>
                <div className="col-3">
                  <Button
                    className="rounded-0 px-5"
                    variant="primary"
                    size="lg"
                    onClick={handleClose}
                  >
                    Close
                  </Button>
                </div>
              </div>
            </form>
          </div>
        </Modal.Body>
      </Modal>
    </React.Fragment>
  );
};

export default CreateInputModal;
