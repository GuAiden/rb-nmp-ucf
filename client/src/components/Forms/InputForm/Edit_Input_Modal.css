.button-clear {
    background-color: transparent;
    border: 2px;
    border-color: white;
}

.button-clear:hover {
    background-color: grey;
    outline: none;
}

.button-clear:focus {
    background-color: grey;
    outline: none;
}

.image-left {
    float: left;
}

.card-button-text {
    font-size: small;
}