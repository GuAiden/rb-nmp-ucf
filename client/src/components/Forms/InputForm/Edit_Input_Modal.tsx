import React, { useState } from 'react';
import { Button, Modal } from 'react-bootstrap';

const EditInputModal: React.FunctionComponent = () => {
  const [show, setShow] = useState(false);

  const handleClose = (): void => {
    setShow(false);
  };

  const handleShow = (): void => {
    setShow(true);
  };
  return (
    <React.Fragment>
      <Button className="rounded-0" variant="primary" onClick={handleShow}>
        Edit
      </Button>

      <Modal show={show} onHide={handleClose} centered size="lg">
        <Modal.Body>
          <p> Yo </p>
        </Modal.Body>
      </Modal>
    </React.Fragment>
  );
};

export default EditInputModal;
