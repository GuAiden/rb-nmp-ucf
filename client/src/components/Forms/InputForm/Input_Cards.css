.card-text {
    color: rgb(136, 136, 136);
    display: flex;
    justify-content: left;
    align-items: flex-start;
    font-size:small;
}

.input-text {
    color: white;
    display: flex;
    justify-content: left;
    align-items: flex-start;
    font-style: italic;
    font-size: large;
    font-weight: 400;
}

.card-body {
    padding-left: 8%;
    background-color: rgb(65, 65, 65);
    border: none;
}

.card-wrapper {
    background-color: rgb(65, 65, 65);
}

.button-clear {
    background-color: transparent;
    border: 2px;
    border-color: white;
}

.button-clear:hover {
    background-color: grey;
    outline: none;
}

.button-clear:focus {
    background-color: grey;
    outline: none;
}

.image-left {
    float: left;
}

.card-button-text {
    font-size: small;
}