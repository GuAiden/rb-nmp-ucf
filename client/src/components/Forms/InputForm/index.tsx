import React, { useState } from 'react';
import CreateInputModal from './createinputmodal';
import gridviewicon from '../../../assets/gridviewicon.png';
import stackedviewicon from '../../../assets/stackedviewicon.png';
import './index.css';
import { Input } from '../inputTypes';

type InputFormProps = {
  onInputChange: (userInputs: Input[]) => void;
};

const InputForm = ({ onInputChange }: InputFormProps): JSX.Element => {
  const [inputList, setInputList] = useState<Input[]>([] as Input[]);

  const handleAddInput = (userInput: Input): void => {
    setInputList(inputList.concat(userInput));
    onInputChange(inputList);
  };

  return (
    <React.Fragment>
      <div className="container bg-dark">
        <div className="container bg-dark pt-4">
          <div className="row justify-content-between">
            <div className="col-4 my-auto">
              <CreateInputModal onAddInput={handleAddInput} />
            </div>
            <div className="col-4 my-auto">
              <button className="icon-wrapper">
                <img
                  src={gridviewicon}
                  alt="gridView"
                  onClick={(): void => console.log('GridView pressed')}
                />
              </button>
              <button className="icon-wrapper">
                <img
                  src={stackedviewicon}
                  alt="stackedView"
                  onClick={(): void => console.log('StackedView pressed')}
                />
              </button>
            </div>
          </div>
          <div className="row">
            {inputList.map((d, idx) => (
              <li className="text-light" key={idx}>
                {d.channelName}
              </li>
            ))}
          </div>
        </div>
      </div>
    </React.Fragment>
  );
};

export default InputForm;
