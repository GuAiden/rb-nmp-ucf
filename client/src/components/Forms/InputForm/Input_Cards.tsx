import React from 'react';
import { Container, Row, Col, Card, Button } from 'react-bootstrap';
import { Input } from '../Input_Types';
import EditInputModal from './Edit_Input_Modal';
import './Input_Cards.css';

type InputCardsProps = {
  inputList: Input[];
};

const InputCards: React.FunctionComponent<InputCardsProps> = ({
  inputList,
}: InputCardsProps) => (
  <Container className="px-5 py-4">
    {/*
    TODO: Find a way to add additional padding without
    nesting containers
    */}
    <Container className="px-5">
      <Row md={3} className="g-5 justify-content-left">
        {inputList.map((_, idx) => (
          <Col>
            <Card className="card-wrapper rounded-0">
              <Card.Body className="card-body">
                <Row>
                  <text className="card-text">Channel Name</text>
                </Row>
                <Row>
                  <text className="input-text">
                    {inputList[idx].channelName}
                  </text>
                </Row>
                <Row className="pt-2">
                  <Col>
                    <text className="card-text">Channel Number</text>
                    <text className="input-text">
                      {inputList[idx].channelNumber}
                    </text>
                  </Col>
                  <Col>
                    <text className="card-text">Units</text>
                    <text className="input-text">{inputList[idx].units}</text>
                  </Col>
                </Row>
                {inputList[idx].conversion ? (
                  <Row className="pt-2">
                    <Col>
                      <text className="card-text">Conversion x:</text>
                      <text className="input-text">{inputList[idx].x}</text>
                    </Col>
                    <Col>
                      <text className="card-text">Conversion y:</text>
                      <text className="input-text">{inputList[idx].y}</text>
                    </Col>
                  </Row>
                ) : (
                  <Row className="py-3 mt-5"></Row>
                )}
                <Row className="justify-content-end g-0 mt-2">
                  <Col md={{ offset: 5 }}>
                    <EditInputModal />
                  </Col>
                  <Col>
                    <Button
                      className="rounded-0 button-clear border px-3 py-0"
                      variant="danger"
                    >
                      <p className="my-auto card-button-text">Delete</p>
                    </Button>
                  </Col>
                </Row>
              </Card.Body>
            </Card>
          </Col>
        ))}
      </Row>
    </Container>
  </Container>
);

export default InputCards;
