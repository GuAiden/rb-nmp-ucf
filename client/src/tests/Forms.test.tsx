import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import InputForm from '../components/Forms/InputForm';

Enzyme.configure({ adapter: new Adapter() });

describe('InputForm component', () => {
  it('renders without failure', () => {
    const handleInputChange = jest.fn();
    const component = shallow(<InputForm onInputChange={handleInputChange} />);
    expect(component).toMatchSnapshot();
  });
});
