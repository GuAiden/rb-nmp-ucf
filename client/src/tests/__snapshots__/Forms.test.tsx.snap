// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm component renders without failure 1`] = `ShallowWrapper {}`;
